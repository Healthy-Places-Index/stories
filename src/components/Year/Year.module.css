.yearLabel{
  font-size: 18px;
  font-weight: bold;
  margin-right: 5px;
}

.yearInput{
  width: 85px;
}

.yearSlider{
  width: calc(100% - 225px);
  display: inline-block;
}