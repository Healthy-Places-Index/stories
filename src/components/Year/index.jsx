/* eslint-disable jsx-a11y/label-has-associated-control */
import React, { useState, useEffect, useRef } from 'react';
import PropTypes from 'prop-types';
import { useQuery, useMutation, gql } from '@apollo/client';
import { Form, Input, Modal, Header, Icon, Button } from 'semantic-ui-react';
import { Slider } from 'react-semantic-ui-range';

import debouncedMutation from '../../lib/debouncedMutation';

import styles from './Year.module.css';

const GET_SLIDE_YEAR = gql`
  query GetSlideYear($slide: ID!) {
    Slide(where: { id: $slide }) {
      id
      year
    }
  }
`;

const UPDATE_SLIDE_YEAR = gql`
  mutation UpdateSlideYear($slide: ID!, $year: Int) {
    updateSlide(id: $slide, data: { year: $year }) {
      id
      year
    }
  }
`;

const Year = ({ slide }) => {
  const { loading, error, data } = useQuery(GET_SLIDE_YEAR, {
    variables: { slide },
  });
  const [onUpdateYear] = useMutation(UPDATE_SLIDE_YEAR);
  const yearTimer = useRef();

  const [tempYear, setTempYear] = useState(null);
  const [inputYear, setInputYear] = useState(1900);
  const [open, setOpen] = useState(false);
  const [rangeError, setRangeError] = useState(false);

  useEffect(() => {
    if (tempYear && data && tempYear !== data.Slide.year) {
      yearTimer.current = debouncedMutation({
        slide,
        timerRef: yearTimer,
        mutation: onUpdateYear,
        values: { year: tempYear },
      });
    }
    setInputYear(tempYear);
  }, [tempYear]);

  useEffect(() => {
    if (data) {
      setTempYear(data.Slide.year);
      setInputYear(data.Slide.year);
    }
  }, [loading, data]);

  const syncYear = () => {
    const newYear = parseInt(inputYear, 10);
    if (newYear && newYear >= 1502 && newYear <= 2021) {
      setRangeError(false);
      setTempYear(newYear);
    } else {
      setRangeError(true);
    }
  };

  return (
    <>
      <Form.Field id="yearInput" inline error={error} disabled={loading}>
        <label className={styles.yearLabel}>Year: </label>
        <Input
          className={styles.yearInput}
          type="number"
          value={inputYear}
          error={rangeError}
          onKeyPress={({ key }) => {
            if (key === 'Enter') syncYear();
          }}
          onChange={(e, { value }) => setInputYear(value)}
          onBlur={syncYear}
        />
        <div className={styles.yearSlider}>
          <Slider
            value={tempYear}
            discrete
            inverted={false}
            settings={{
              value: tempYear,
              min: 1502,
              max: 2021,
              step: 1,
              onChange: setTempYear,
            }}
          />
        </div>
      </Form.Field>
      <Modal
        basic
        onClose={() => setOpen(false)}
        onOpen={() => setOpen(true)}
        open={open}
        size="small"
      >
        <Header icon>
          <Icon name="calendar times outline" />
          Are you sure you want to change the year?
        </Header>
        <Modal.Content>
          <p>{`You have a selected feature or basemap selected that is tied to ${tempYear}. Changing the year will remove these selections.`}</p>
        </Modal.Content>
        <Modal.Actions>
          <Button basic color="red" inverted onClick={() => setOpen(false)}>
            <Icon name="remove" />
            <span>{`Keep ${tempYear}`}</span>
          </Button>
          <Button
            negative
            inverted
            onClick={() => {
              setOpen(false);
            }}
          >
            <Icon name="trash" />
            <span>Clear selection and change year</span>
          </Button>
        </Modal.Actions>
      </Modal>
    </>
  );
};

Year.propTypes = {
  slide: PropTypes.string.isRequired,
};

export default Year;
