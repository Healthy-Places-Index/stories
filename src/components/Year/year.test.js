import React from 'react';
import renderer, { act } from 'react-test-renderer';
import { MockedProvider } from '@apollo/client/testing';

import Year, { GET_SLIDE_YEAR } from './index';

const mocks = [
  {
    request: {
      query: GET_SLIDE_YEAR,
      variables: {
        slide: '1',
      },
    },
    result: {
      data: {
        Slide: {
          id: '1',
          year: 1900,
        },
      },
    },
  },
];

jest.mock('react-semantic-ui-range', () => {
  const Slider = () => <div className="slider" />;
  return { Slider };
});

jest.mock('semantic-ui-react/dist/commonjs/addons/Portal/Portal', () => ({ children }) => children);

describe('Year', () => {
  it('matches snapshot', async () => {
    const component = renderer.create(
      <MockedProvider mocks={mocks} addTypename={false}>
        <Year slide="1" />
      </MockedProvider>
    );

    await act(() => new Promise(resolve => setTimeout(resolve, 0)));
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
