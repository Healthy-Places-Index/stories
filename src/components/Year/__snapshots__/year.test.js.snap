// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Year matches snapshot 1`] = `
<div
  className="inline field"
  id="yearInput"
>
  <label
    className="yearLabel"
  >
    Year: 
  </label>
  <div
    className="ui input yearInput"
  >
    <input
      onBlur={[Function]}
      onChange={[Function]}
      onKeyPress={[Function]}
      type="number"
      value={1900}
    />
  </div>
  <div
    className="yearSlider"
  >
    <div
      className="slider"
    />
  </div>
</div>
`;
